streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -38,346 +40,644 @@
+                    throw new IllegalStateException("The corresponding changelog restorer for " + partition +
+                        " does not exist, this should not happen.");
+                }
+                if (changelogMetadata.changelogState != ChangelogState.RESTORING) {
+                    throw new IllegalStateException("The corresponding changelog restorer for " + partition +
+                        " has already transited to completed state, this should not happen.");
+                }
+
+                final List<ConsumerRecord<byte[], byte[]>> records = polledRecords.records(partition);
+                final Iterator<ConsumerRecord<byte[], byte[]>> iterator = records.iterator();
+                while (iterator.hasNext()) {
+                    final ConsumerRecord<byte[], byte[]> record = iterator.next();
+                    if (record.key() == null) {
+                        log.warn("Read changelog record with null key from changelog {} at offset {}, " +
+                            "skipping it for restoration", changelogMetadata.changelogPartition, record.offset());
+                        iterator.remove();
+                    }
+                }
+
+                // TODO: we always try to restore as a batch when some records are accumulated, which may result in
+                //       small batches; this can be optimized in the future, e.g. wait longer for larger batches.
+                if (!records.isEmpty()) {
+                    if (changelogMetadata.bufferedRecords != null && !changelogMetadata.bufferedRecords.isEmpty()) {
+                        changelogMetadata.bufferedRecords.addAll(records);
+                    } else {
+                        changelogMetadata.bufferedRecords = records;
+                    }
+
+                    restoreChangelog(changelogMetadata);
+                }
             }
-        });
-
-        final Iterator<TopicPartition> iter = initializable.iterator();
-        while (iter.hasNext()) {
-            final TopicPartition topicPartition = iter.next();
-            final Long restoreOffset = restoreToOffsets.get(topicPartition);
-            final StateRestorer restorer = stateRestorers.get(topicPartition);
-
-            if (restorer.checkpoint() >= restoreOffset) {
-                restorer.setRestoredOffset(restorer.checkpoint());
-                iter.remove();
-                completedRestorers.add(topicPartition);
-            } else if (restoreOffset == 0) {
-                restorer.setRestoredOffset(0);
-                iter.remove();
-                completedRestorers.add(topicPartition);
+        }
+    }
+
+    /**
+     * returns a flag whether offset limit caused not all records restored
+     */
+    private void restoreChangelog(final ChangelogMetadata changelogMetadata) {
+        final TopicPartition partition = changelogMetadata.changelogPartition;
+        final int numRecords = IntStream.range(0, changelogMetadata.bufferedRecords.size())
```

Comment:
I feel like a simple for loop would be much easier to read, but if others don't agree I'm fine with this if we can maybe improve the variable name. eg numRecords -> numRecordsBeforeLastOffset (or something better but to that effect)

Reply:
The tricky part is that the underlying `restoreCallback.restoreBatch` inside `stateManager#restore` would try to restore all the passed in records in a list, so I have to pass in a sub-list if the buffered records exceed the end / limit offset. And also after that we want to clear the sub-list from the buffered records in a most efficient way (that's why I used an ArrayList).

Reply:
@ableegoldman After some thoughts I refactored the code to record the "limit index" in the buffered records while checking null-keys and then use that limit to get the sublist directly.

Also did some minor optimizations to save array-shift / copy from the fetched records.

Reply:
I tried to avoid array-copy from buffered records to the records to restore list, but I agree it did look a bit weird with that optimization. Does the current refactored code look clearer to you folks?